# Современные методы анализа данных и машинного обучения, БИ

## НИУ ВШЭ, 2025-26 учебный год

## Групповой проект №5. Deep Learning

## Общая информация

__Дата выдачи:__ 25.05.2026

__Дедлайн по записи в таблицу:__ 4:00 03.06.2026

__Дедлайн по загрузке работ:__ 4:00 15.06.2026

__Защита проекта:__ ~

## Формат выполнения

Групповой проект представляет собой творческое инициативное групповое задание, которое выполняется студентами в командах от 3 до 6 человек. В то же время следует помнить, что чем больше в команде людей, тем выше предъявляются требования в целом к команде, а также к выполняемому ею проекту. Команды разрешается формировать из студентов разных групп.

При оценивании результатов выполнения проекта командой в обязательном порядке проверяется и учитывается в оценке вклад **каждого участника** в общую работу команды. Вследствие этого крайне важно ответственно подойти к выбору сокомандников, распределению обязанностей внутри команды, а также пресечению наличия "мертвых душ" в команде.

При необходимости команда большинством голосов имеет право исключить человека из команды **до загрузки работ** в систему [Anytask](https://anytask.org/), сообщив об этом преподавателю, в случае если такой человек никак не проявлялся в ходе выполнения проекта и являлся как раз таки той самой "мертвой душой". Помните про эту возможность, поскольку наличие "мертвых душ" может негативно повлиять на качество и оценку для всего проекта и, как следствие, в целом для всей команды!

Запись составов команд осуществляется по ссылке в таблице:

https://docs.google.com/spreadsheets/d/16IB8mURLGKoipq0Jld6wOGPMpJbRQ8GCkSZKbIP-XYA/edit?usp=sharing

## Отправка материалов группового проекта

Во избежание использования при выполнении группового проекта генеративных моделей (ChatGPT и так далее) за рамками справочной и образовательной информации (то есть, например, для автоматической генерации кода по одному из этапов) **одному из участников команды** необходимо предварительно загрузить работу в систему [Anytask](https://anytask.org/). Наличие сгенерированного кода или любого другого случая "несанкционированного" использования нейросетей будет считаться плагиатом, и такой проект будет оценен в 0 баллов для всех участников. При отсутствии своевременно загруженных файлов с проектом команде также выставляется 0 за проект.

*Примечание: презентацию проекта загружать необязательно, важно отправить содержательные файлы с кодом и ссылку на репозиторий GitHub.*

## Оценивание и штрафы



Групповой проект оценивается исходя из устной защиты (презентации) командой выполненного ими проекта перед комиссией. Комиссия составляется из преподавателей курса, ассистентов курса, а также приглашенных экспертов из индустрии.

В рамках каждой защиты комиссия выставляет два типа оценок.

Первая оценка — это оценка для всего проекта в целом. Она ставится комиссией, исходя из глубины проработки командой предметной области, сложности применяемых в проекте инструментов, а также прочих аспектов, связанных с содержанием выполняемого в проекте ТЗ. При выставлении этой оценки комиссия ориентируется на критерии, более подробно представленные ниже в разделе "Критерии оценки проекта".

Вторая оценка — это индивидуальная оценка каждого участника команды. Именно эта оценка выставляется впоследствии в ведомость. Индивидуальная оценка участника команды определяется комиссией, исходя из того, насколько участник ориентируется в выполненном командой проекте, как он отвечает на адресованные персонально ему вопросы комиссии, насколько в принципе прослеживается его роль в проекте, а также исходя из других аспектов, связанных с участием конкретного человека в деятельности команды. Индивидуальная оценка не может превосходить оценку команды за проект в целом (первую оценку), однако, в случае когда участник спокойно и корректно отвечает на все адресованные ему вопросы, свободно ориентируется в материале и содержании выполненного проекта, а также когда его роль в работе команды не вызывает сомнений, — в таком случае, индивидуальная оценка участника становится равной оценке для всего проекта (т.е. в идеальном случае — вторая оценка равна первой оценке). Кроме того, в исключительных случаях, при демонстрации участником на защите превосходящих ожидания уровней понимания, участия и вовлечения в рамках проекта, комиссия может своим решением повысить индивидуальную оценку для этого участника на 0,5 или 1 балл относительно общей оценки за проект (первой оценки).

Во избежание субъективности и/или предвзятости оценки тех или иных членов комиссии в отношении тех или иных команд, а также каких-либо иных проявлений неравных условий — распределение команд между комиссиями осуществляется абсолютно рандомно с использованием средств библиотеки Numpy :)) — в качестве рэндом сида выступает дата, в которую производится распределение, например, `np.random.seed(25052026)`.

Обратите, пожалуйста, внимание на то, что бОльшая часть комиссий осуществляет приём проектов в очной форме, однако могут присутствовать и 1-2 комиссии, которые будут работать дистанционно с использованием средств видеосвязи. В случае попадания команды путем случайного распределения в комиссию, работающую в очной форме, *не допускается* какое-либо изменение формата проведения для команды на онлайн-режим, равно как и любые другие перемещения команд между комиссиями (смены комиссий), для сохранения честности и непредвзятости случайного распределения.

Явка студента на защиту проекта является обязательной; вклад студента без его присутствия не оценивается и автоматически является равным нулю. Для дистанционных комиссий дополнительным требованием является обязательное наличие включенной камеры в течение всего времени нахождения в конференции

## Описание задания

В современной бизнес-индустрии использование глубинного обучения приобретает критическое значение. Компании активно внедряют нейронные сети для решения самых разных задач: от анализа структурированных табличных данных до работы с изображениями и текстами. Это позволяет бизнесу принимать более обоснованные решения и оперативно реагировать на изменения рынка.

В данном проекте вы на реальных данных попрактикуетесь в построении и сравнении нейросетевых архитектур для решения двух бизнес-задач разной природы. Первая задача будет связана с табличными данными и решается с помощью полносвязных нейронных сетей. Вторая - с неструктурированными данными: изображениями или текстами, - и решается с помощью более сложных архитектур глубинного обучения. Помимо этого, вы получите опыт организации работы над ML-проектом так, как это принято в индустрии: с версионированием кода, воспроизводимыми экспериментами и сохранением всех артефактов.

Принципиально важно, что обе бизнес-задачи должны быть **содержательно связаны между собой** и **относиться к одной и той же компании или платформе** — таким образом, чтобы вместе они образовывали единое и логически целостное решение для выбранного бизнес-контекста. Проект, в котором две задачи решаются для разных компаний или никак не связаны между собой, оценивается в **0 баллов** для всех участников команды.

### Этапы проекта

Ниже вашему вниманию предлагаются примерные этапы реализации проекта:

- Выбор компании и/или платформы;
- Формулировка двух связанных бизнес-задач, которые могла бы хотеть решить компания/платформа на основе данных: одна на табличных данных, другая на неструктурированных (изображения или текст);
- Обоснование сформулированных бизнес-задач и описание того, как предлагаемое решение вписывается в бизнес-процессы компании;
- Получение необходимых данных путём поиска открытых источников/датасетов, парсинга/работы с API;
- Формулировка задач в терминах глубинного обучения для решения поставленных бизнес-задач;
- Выбор и обоснование архитектур нейронных сетей для каждой из задач;
- Первичная обработка данных;
- Построение, обучение и сравнение нейросетевых архитектур с обязательным логированием экспериментов;
- Получение и агрегация результатов, финальные выводы.

### Требования к проекту

В качестве результата выполнения группового проекта ваша команда должна получить презентацию и защитить её перед комиссией, а также предоставить ссылку на репозиторий GitHub с кодом.

Оформление презентации остаётся полностью на ваше усмотрение, но помните, что результат должен быть релевантен для демонстрации бизнес-заказчику — комиссию, принимающую вашу работу, правильнее всего воспринимать именно в таком качестве. Например, вставлять в презентацию строчки кода или злоупотреблять скринами блокнота не рекомендуется.

С точки зрения концепции выполнения проекта, вам необходимо принять на себя роль дата-инженеров: продумать идеологию получения и извлечения данных; дата-аналитиков: соотнести бизнес-задачу с формулировками на языке данных; и главное, безусловно, — дата-сайентистов: реализовать поставленные задачи с помощью методов глубинного обучения и извлечь полезные инсайты из проделанной работы. А в конце, конечно, презентовать всю эту красоту в понятном виде потенциальному заказчику.

Разумеется, чем сложнее и чем нетривиальнее будет ваша задача, а также чем глубже и чем осмысленнее будут выводы, извлечённые в рамках этой задачи, — тем выше окажется качество вашего проекта и тем выше он будет оценён!


### Полезные замечания и советы

- Графики и визуализации — это очень хорошо — все любят графики
- Следите за тем, чтобы все графики на слайдах можно было прочитать даже без ваших комментариев
- Не нужно делать просто текстовые слайды
- Стройте свой проект исходя из того, что ваша цель — представить полезную информацию для бизнеса
- Любой хороший Data Scientist должен уметь работать в команде, поэтому распределение ваших усилий в проекте остается за вами


## Критерии оценки проекта

Оценка за групповой проект в целом — то есть первая из двух оценок — выставляется по 10-балльной шкале, исходя из 5 критериев. Также комиссия на своё усмотрение может добавить (но не снизить) некоторое количество бонусных баллов за рамками данных критериев.

Каждый из критериев будет перечислен и рассмотрен ниже.

### Критерий 1. Выбор задач и бизнес-обоснование (до 2 баллов)

Данный критерий оценивает, насколько обоснованными и осмысленными и **содержательно связанными** получились две решаемые в проекте бизнес-задачи; насколько полно был осуществлён переход от бизнес-задач к задачам глубинного обучения (насколько в реальности постановка задач соответствует исходной постановке задачи с точки зрения интересов компании); а также насколько корректными, объёмными и прикладными для бизнеса получились выявленные в ходе исследования инсайты.

По данному критерию можно получить **максимальный балл** только при условии, что в презентации представлена схема, иллюстрирующая, как предлагаемое решение вписывается в бизнес-процессы компании: откуда поступают данные, каким образом они обрабатываются, как модель генерирует предсказания и каким образом эти предсказания используются для принятия бизнес-решений. Строгая нотация не обязательна — важна логика и понятность схемы для стороннего наблюдателя.

### Критерий 2. Инструменты извлечения и обработки данных (до 1,5 баллов)

Данный критерий оценивает техническую реализацию процесса сбора данных ручным и автоматизированным образом, использованные средства предобработки, подготовки данных и EDA. В рамках этого критерия будет рассматриваться спектр различных технических инструментов, внедрённых в проект; корректность, эффективность и правильность их применения; а также сложность и новизна алгоритмов, подходов и библиотек, применяемых к данным.

### Критерий 3. Глубинное обучение на табличных данных (до 2 баллов)

Данный критерий оценивает применение полносвязных нейронных сетей для решения бизнес-задачи на табличных данных. В рамках этого критерия будет рассматриваться корректность постановки задачи, осмысленность выбора архитектуры, качество предобработки данных, а также глубина сравнительного анализа различных конфигураций сети.

*__Важно:__ в случае **оценки 0** за этот критерий оценка за весь проект становится равна оценке за этот критерий.*

*По данному критерию можно получить **отличное от нуля количество баллов** только при условии корректного применения полносвязной нейронной сети для решения поставленной задачи на табличных данных.*

*По данному критерию можно получить **1 балл и выше** только при условии осуществления сравнительного анализа различных конфигураций и сложности архитектуры: должно быть рассмотрено не менее трёх вариантов, значимо отличающихся друг от друга — например, по глубине сети, размеру скрытых слоёв, функциям активации, наличию регуляризации. Использование каждой конфигурации, как и подбор гиперпараметров, должны быть подробно описаны и обоснованы; в рамках сравнения должны быть получены и сформулированы явные, значимые выводы о том, как изменение архитектуры влияет на качество решения задачи.*

*По данному критерию можно получить **максимальный балл в 2 балла** только при условии корректного и обоснованного применения архитектурных решений или подходов, не рассматривавшихся в рамках соответствующих лекций и семинаров. Кроме того, **максимальный балл в 2 балла** по данному критерию недостижим без логирования экспериментов в соответствии с требованиями **Критерия 5**. При этом ожидается, что каждая из сравниваемых конфигураций залогирована отдельным запуском с сохранением параметров архитектуры и обучения, метрик качества, а также артефакта обученной модели, позволяющего воспроизвести её предсказания без повторного обучения.*

### Критерий 4. Глубинное обучение на неструктурированных данных  (до 3 баллов)

Данный критерий оценивает применение методов глубинного обучения для решения бизнес-задачи на неструктурированных данных: изображениях или текстах. В рамках этого критерия будет рассматриваться спектр различных технических инструментов, внедрённых в проект; корректность, эффективность и правильность их применения; а также сложность и новизна алгоритмов, подходов и библиотек, применяемых к данным. Кроме того, будет оценен сравнительный анализ архитектур нейронных сетей, использованных в решении.

*__Важно:__ в случае **оценки 0** за этот критерий оценка за весь проект становится равна оценке за этот критерий.*

*По данному критерию можно получить **отличное от нуля количество баллов** только при условии использования в проекте сверточных нейронных сетей, рекуррентных нейронных сетей или же более сложных архитектур глубоких нейронных сетей.*

*По данному критерию можно получить **2 балла** только при условии осуществления сравнительного анализа эффективности различных архитектур нейронных сетей внутри вашего проекта на одних и тех же данных. Самих архитектур должно быть не менее трёх; они должны значимо и осмысленно друг от друга отличаться; их использование, как и настройка гиперпараметров, должны быть подробно описаны и обоснованы. Кроме того, в рамках сравнительного анализа должны быть получены и сформулированы явные, значимые выводы о сравнительной эффективности той или иной архитектуры.*

*По данному критерию можно получить **максимальный балл в 3 балла** только при условии корректного и обоснованного применения инструментов/подходов/библиотек для построения моделей, не рассматривавшихся в рамках соответствующих лекций и семинаров. Кроме того, **максимальный балл в 3 балла** по данному критерию недостижим без логирования экспериментов в соответствии с требованиями **Критерия 5**. При этом ожидается, что каждая из сравниваемых архитектур залогирована отдельным запуском с сохранением параметров архитектуры и обучения, метрик качества в разрезе эпох, а также артефакта обученной модели, позволяющего воспроизвести её предсказания без повторного обучения.*

### Критерий 5. Инструменты промышленной разработки (до 1,5 баллов)

Данный критерий оценивает, насколько грамотно и профессионально команда организовала процесс совместной разработки и обеспечила воспроизводимость своих экспериментов. В рамках этого критерия рассматриваются две составляющие: оформление репозитория GitHub и логирование экспериментов.

*По данному критерию можно получить **0,5 балла** при выполнении всех следующих требований к репозиторию GitHub:*

- наличие файла `README.md` с описанием проекта, используемых данных и структуры репозитория — после прочтения этого файла у стороннего человека не должно остаться вопросов о том, что делает проект и как в нём разобраться;
- наличие файла `.gitignore`;
- осмысленная структура репозитория: код организован по директориям, понятно, где лежат данные, где модели, где ноутбуки;
- коммиты делаются осмысленно и регулярно по ходу работы над проектом, с понятными комментариями — по истории репозитория должно быть видно, что команда работала совместно на протяжении всего времени, а не залила всё в одном коммите накануне дедлайна.

*По данному критерию можно получить **1 балл** при дополнительном выполнении следующего условия: все эксперименты с моделями — как на табличных, так и на неструктурированных данных — залогированы с использованием специализированного инструмента логирования экспериментов (например, MLflow, Weights & Biases, ClearML, DagsHub или аналогичного). Каждый эксперимент должен быть зафиксирован отдельным запуском и содержать: параметры модели и обучения, метрики качества — для табличных задач в целом по выборке, для задач на неструктурированных данных дополнительно в разрезе эпох, — а также артефакт обученной модели, позволяющий воспроизвести её предсказания без повторного обучения. Хранение файлов моделей исключительно напрямую в репозитории без привязки к системе логирования не засчитывается.*

*По данному критерию можно получить **1,5 балла** при дополнительном выполнении следующего условия: в системе логирования сохранены также данные, использованные при обучении и валидации моделей, — таким образом, чтобы любой эксперимент можно было полностью воспроизвести, имея доступ только к системе логирования, без необходимости искать данные отдельно.*

#Бизнес составляющая проекта

**Компания:** Медицинская клиника

**Бизнес задача 1:**
Атоматизация скрининга пациента для проверки наличия пневмонии на основе симптомов

**Бизнес задача 2:**
Автоматизация проверки рентгеновских снимков грудной клетки пациента для выявления пневмонии


**Обоснование бизнес задач:**
В период сезонных заболеваний в поликлиники обращается множество пациентов с признаками респираторных заболеваний. Для быстрого выявления тяжелого заболевания - пневмонии, необходима автоматизация этого процесса.

**Решение:**
Внедрение двухэтапной системы принятия решения:
1. Модель оценивает вероятность наличия пневмонии по клиническим симптомам для необходимости прохождения рентгена легких
2. Модель оценивает рентгеновский снимок грудной клетки пациента для помощи определения наличия пневмонии

**Связь задач:**
Первая задача используется для первичного скрининга пациента, вторая после прохождения рентгена. В совокупности они образуют полный цикл выявления пневмонии у пациента

**Польза:**
- Ускорение процесса диагностики
- Сокращение нагрузки на медицинский персонал
- Быстрое выявление пациентов с пневмонией
- Более точное назначение рентгена, за счет чего идет уменьшение вредного влияния этого на больных


**Схема работы в процессах компании:**
Пациент обращается в поликлинику -> ему отправляется специальная форма, в которой предлагается заполнить данные о его симптомах -> оценка вероятности наличия пневмонии -> назначение рентгена -> анализ снимка -> при подтверждении этого заболевания - приоритетная запись на следующий прием к врачу -> итоговое решение врача и назначение лечения



## Работа с датасетом табличных данных

Скачать датасет можно по ссылке https://www.kaggle.com/datasets/essienmary/pneumonia-dataset

In [1]:
#импортируем библиотеки
import pandas as pd
import numpy as np

In [2]:
data1 = pd.read_csv('pneumonia_dataset.csv')

#### Выполним предобработку первого датасета (табличные данные)

In [3]:
#посмотрим на типы данных в колонках
data1.dtypes

,0
PatientID,object
Gender,object
Age,int64
Cough,object
Fever,object
Shortness_of_breath,object
Chest_pain,object
Fatigue,object
Confusion,object
Oxygen_saturation,object


In [4]:
#проверим на пропуски
data1.isna().sum()

,0
PatientID,0
Gender,0
Age,0
Cough,312
Fever,310
Shortness_of_breath,317
Chest_pain,343
Fatigue,105
Confusion,0
Oxygen_saturation,0


In [5]:
#посмотрим на данные
data1.head(5)

,PatientID,Gender,Age,Cough,Fever,Shortness_of_breath,Chest_pain,Fatigue,Confusion,Oxygen_saturation,Crackles,Xray,WBC_count,Sputum_color,Temperature,Diagnosis
0,PNEU4091,M,3,Bloody,High,Mild,Mild,NaN,No,92,Yes,-,-,-,39.7,Yes
1,PNEU6231,M,3,Dry,NaN,NaN,NaN,Mild,-,-,Yes,-,-,Clear,37.0,No
2,PNEU6486,M,3,Dry,NaN,NaN,NaN,Mild,No,99,No,-,-,-,38.0,No
3,PNEU7601,F,3,NaN,Low,Severe,NaN,Moderate,No,99,Yes,-,-,Clear,40.9,Yes
4,PNEU8471,F,3,NaN,High,Moderate,Severe,Mild,-,91,No,-,-,Green,38.1,Yes


#### Видим, что у нас в датасете практически все столбцы содержат данные типа object, поэтому преобразуем их

In [6]:
print(data1['Gender'].unique())
#в столбце Gender нет пропусков, при этом все данные состоят только из F и M,
#поэтому преобразуем их в формат 0 и 1
data1['Gender'] = data1['Gender'].apply(lambda x: 1 if x == 'M' else 0)
print(data1['Gender'].unique())

['M' 'F']
[1 0]


Во многих столбцах данные категориальные, при этом уникальных значений в них не много. Так как не понятно отношение порядка между признаками, применим к данным OneHotEncoding, а не LabelEncoding или OrdinalEncoding.

Это позволит при обучении получить большую точность, так как у модели будет более полная информация о признаках, ведь на моменте применения LabelEncoding эта информация могла потеряться, так как модель могла бы считать влияние разницы между сухим и влажным таким же, как влиение разницы между влажным кашлем и кашлем с кровью, что может на самом деле быть не так

In [7]:
#доументация https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html
from sklearn.preprocessing import OneHotEncoder
categorial = ['Cough', 'Fever', 'Shortness_of_breath', 'Chest_pain', 'Fatigue', 'Crackles', 'Sputum_color']

encoder = OneHotEncoder(drop='first', sparse_output=False)
encoded = encoder.fit_transform(data1[categorial])
encoded_cols = encoder.get_feature_names_out(categorial)
encoded_df = pd.DataFrame(encoded, columns=encoded_cols, index=data1.index)

data1 = data1.drop(columns=categorial)
data1 = pd.concat([data1, encoded_df], axis=1)

In [8]:
data1.head(5)

,PatientID,Gender,Age,Confusion,Oxygen_saturation,Xray,WBC_count,Temperature,Diagnosis,Cough_Dry,...,Chest_pain_Severe,Chest_pain_nan,Fatigue_Moderate,Fatigue_Severe,Fatigue_nan,Crackles_Yes,Sputum_color_Bloody,Sputum_color_Clear,Sputum_color_Green,Sputum_color_Yellow
0,PNEU4091,1,3,No,92,-,-,39.7,Yes,0.0,...,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0
1,PNEU6231,1,3,-,-,-,-,37.0,No,1.0,...,0.0,1.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0
2,PNEU6486,1,3,No,99,-,-,38.0,No,1.0,...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,PNEU7601,0,3,No,99,-,-,40.9,Yes,0.0,...,0.0,1.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0
4,PNEU8471,0,3,-,91,-,-,38.1,Yes,0.0,...,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0


In [9]:
data1.columns

Index(['PatientID', 'Gender', 'Age', 'Confusion', 'Oxygen_saturation', 'Xray',
       'WBC_count', 'Temperature', 'Diagnosis', 'Cough_Dry', 'Cough_Wet',
       'Cough_nan', 'Fever_Low', 'Fever_Moderate', 'Fever_nan',
       'Shortness_of_breath_Moderate', 'Shortness_of_breath_Severe',
       'Shortness_of_breath_nan', 'Chest_pain_Moderate', 'Chest_pain_Severe',
       'Chest_pain_nan', 'Fatigue_Moderate', 'Fatigue_Severe', 'Fatigue_nan',
       'Crackles_Yes', 'Sputum_color_Bloody', 'Sputum_color_Clear',
       'Sputum_color_Green', 'Sputum_color_Yellow'],
      dtype='object')

In [10]:
#так как мы хотим определять наличие/отсутствие пневмонии по опросу симптомов,
#удалим из датасета признаки, для которых необходимо сдавать анализы, например уровень кислорода
#или количество лейкоцитов
#в том числе удалим и id клиента, так как это-не значимый признак для определения наличия пневмонии
#также удалим признак Confusion, так как не понятна его физическая составляющая
data1 = data1.drop(columns = ['PatientID', 'Confusion', 'Xray', 'WBC_count', 'Oxygen_saturation'])

In [11]:
#преобразуем целевую переменную - diagnosis
data1['Diagnosis'] = data1['Diagnosis'].apply(lambda x: 1 if x == 'Yes' else 0)

In [12]:
#смотрим на баланс классов в целевой переменной
data1['Diagnosis'].value_counts()

,count
Diagnosis,
1,408
0,302


Видим, что классы не сбалансированны, хотя разница не сильно велика

In [13]:
#посмотрим на корреляцию признаков с целевой переменной - Diagnosis
correlations = data1.corr()['Diagnosis'].abs().sort_values(ascending=False)
print(correlations)

Diagnosis                       1.000000
Fever_nan                       0.512024
Chest_pain_nan                  0.502275
Cough_nan                       0.432160
Shortness_of_breath_nan         0.430730
Sputum_color_Clear              0.415674
Cough_Wet                       0.374305
Chest_pain_Severe               0.370358
Fatigue_Severe                  0.362415
Sputum_color_Green              0.360419
Fatigue_nan                     0.358418
Fever_Moderate                  0.356413
Shortness_of_breath_Moderate    0.356413
Shortness_of_breath_Severe      0.348344
Fatigue_Moderate                0.346314
Chest_pain_Moderate             0.344279
Sputum_color_Bloody             0.336083
Temperature                     0.220723
Crackles_Yes                    0.202253
Age                             0.166676
Cough_Dry                       0.096436
Gender                          0.070685
Sputum_color_Yellow             0.057968
Fever_Low                       0.017779
Name: Diagnosis,

### Работа с полносвязными сетями для табличных данных

In [14]:
!pip install wandb

In [15]:
#импорт библиотек

import os
import wandb # для логирования

import numpy as np
import random
from tqdm import *

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim # для оптимизаторов
from torchvision import datasets # для данных
import torchvision.transforms as transforms # для преобразований тензоров

import matplotlib.pyplot as plt

In [16]:
# Зафиксируем seed для воспроизводимости

def seed_everything(seed):
    random.seed(seed) # фиксируем генератор случайных чисел
    os.environ['PYTHONHASHSEED'] = str(seed) # фиксируем заполнения хешей
    np.random.seed(seed) # фиксируем генератор случайных чисел numpy
    torch.manual_seed(seed) # фиксируем генератор случайных чисел pytorch
    torch.cuda.manual_seed(seed) # фиксируем генератор случайных чисел для GPU
    #torch.backends.cudnn.deterministic = True # выбираем только детерминированные алгоритмы (для сверток)
    #torch.backends.cudnn.benchmark = False # фиксируем алгоритм вычисления сверток

In [17]:
class CFG:

# Задаем параметры нашего эксперимента

  api = "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"# вписать свой API Wandb
  project = "gp5-project"# вписать название эксперимента, который предварительно надо создать в Wandb
  entity = None # ввести свой логин
  num_epochs = 10 # количество эпох
  train_batch_size = 64 # размер батча обучающей выборки
  test_batch_size = 256 # размер батча тестовой выборки
  num_workers = 2 # количество активных процессов на загрузку данных
  lr = 0.001 # learning_rate
  seed = 2022 # для функции воспроизводимости
  wandb = True # флаг использования Wandb

In [18]:
# функция перевода класса конфигурации в словарь

def class2dict(f):
  return dict((name, getattr(f, name)) for name in dir(f) if not name.startswith('__'))

In [19]:
# будем загружать данные, сразу преобразуя их в тензор

transform = transforms.Compose(
    [transforms.ToTensor()])

In [20]:
from sklearn.model_selection import train_test_split
from torch.utils.data import TensorDataset, DataLoader
from sklearn.preprocessing import StandardScaler


X = data1.drop(columns=['Diagnosis'])
y = data1['Diagnosis']

#нормируем при помощи StandartScaller
scaler = StandardScaler()
columns = X.columns
X = scaler.fit_transform(X)
X = pd.DataFrame(X, columns=columns)

#разбиваем на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

#конвертируем в тензор
X_train_t = torch.tensor(X_train.values, dtype=torch.float32)
y_train_t = torch.tensor(y_train.values, dtype=torch.long)
X_test_t = torch.tensor(X_test.values, dtype=torch.float32)
y_test_t = torch.tensor(y_test.values, dtype=torch.long)

train_dataset = TensorDataset(X_train_t, y_train_t)
test_dataset = TensorDataset(X_test_t, y_test_t)

train_loader = DataLoader(train_dataset, batch_size=CFG.train_batch_size,
                          shuffle=True, num_workers=CFG.num_workers, pin_memory=True)
test_loader = DataLoader(test_dataset, batch_size=CFG.test_batch_size,
                         shuffle=False, num_workers=CFG.num_workers, pin_memory=True)

In [21]:
examples = enumerate(test_loader)
batch_idx, (example_data, example_targets) = next(examples) # итерируемся по тестовым данным

example_data.shape # проверим размерность

/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:1118: UserWarning: 'pin_memory' argument is set as true but no accelerator is found, then device pinned memory won't be used.
  super().__init__(loader)


torch.Size([142, 23])

In [22]:
#строим первую модель
class first_net(nn.Module): # наследуемся от класса nn.Module
    def __init__(self):
        super(first_net,self).__init__()
        hidden_1 = 256
        hidden_2 = 128

        self.net = torch.nn.Sequential(
                      torch.nn.Linear(23, hidden_1),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_1, hidden_2),
                      torch.nn.Linear(hidden_2, 2)
                    )

    def forward(self,x):
        x = self.net(x)
        return x

In [23]:
model1 = first_net()
#model = model.cuda() # переводим модель на GPU
print(model1) # посмотрим на нашу модель

first_net(
  (net): Sequential(
    (0): Linear(in_features=23, out_features=256, bias=True)
    (1): ReLU()
    (2): Linear(in_features=256, out_features=128, bias=True)
    (3): Linear(in_features=128, out_features=2, bias=True)
  )
)


In [24]:
# функция потерь (cross-entropy)
criterion = nn.CrossEntropyLoss()

# SGD (классический) со скоростью обучения (learning rate) = 0.01
optimizer = torch.optim.SGD(model1.parameters(),lr = 0.01)

In [24]:
# функция обучения модели
def train(model, device, train_loader, optimizer, criterion, epoch, WANDB):
    model.train() # обязательно переводим в режим обучения
    test_loss = 0
    correct = 0

    n_ex = len(train_loader)

    for batch_idx, (data, target) in tqdm(enumerate(train_loader), total=n_ex):
        data, target = data.to(device), target.to(device) # переводим картинки и таргеты на GPU
        # обнуляем градиенты!
        optimizer.zero_grad()
        # прямой проход
        output = model(data)
        pred = output.argmax(dim=1, keepdim=True)
        correct += pred.eq(target.view_as(pred)).sum().item() # определяем корректность предсказания
        train_loss = criterion(output, target) # считаем значение функции потерь
        # обратный проход
        train_loss.backward()
        # делаем градиентный шаг оптимизатором
        optimizer.step()
        # считаем метрики и лосс
    tqdm.write('\nTrain set: Average loss: {:.4f}, Accuracy: {:.0f}%'.format(
        train_loss, 100. * correct / len(train_loader.dataset)))

    if WANDB:
        wandb.log({'train_loss': train_loss,
                   'train_accuracy': correct / len(train_loader.dataset)})

In [25]:
# функция тестирования
def test(model, device, test_loader, criterion, WANDB):
    model.eval() # переводем модель в режим инференса
    test_loss = 0
    correct = 0
    # показываем, что обученич нет и градиенты не обновляются
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss = criterion(output, target) # считаем значение функции потерь
            pred = output.argmax(dim=1, keepdim=True)
            correct += pred.eq(target.view_as(pred)).sum().item()
            # считаем метрики
    tqdm.write('Test set: Average loss: {:.4f}, Accuracy: {:.0f}%'.format(
        test_loss, 100. * correct / len(test_loader.dataset)))

    if WANDB:
        wandb.log({'test_loss': test_loss,
                   'test_accuracy': correct / len(test_loader.dataset)})

In [26]:
# основная функция для экспериментов
def main(model):
    # настраиваем окружение для wandb
    if CFG.wandb:
        os.environ["WANDB_API_KEY"] = CFG.api
        wandb.init(project=CFG.project, entity=CFG.entity, reinit=True, config=class2dict(CFG))

    use_cuda = torch.cuda.is_available() # проверяем доступность GPU

    seed_everything(CFG.seed) # фиксируем сиды

    device = torch.device("cuda" if use_cuda else "cpu") # выделили устройство

    if CFG.wandb:
        wandb.watch(model, log='all') # логируем все (метрики, лоссы, градиенты)

    for epoch in range(1, CFG.num_epochs + 1): # цикл на эпохи
        print('\nEpoch:', epoch)
        train(model, device, train_loader, optimizer, criterion, epoch, CFG.wandb)
        test(model, device, test_loader, criterion, CFG.wandb)
    print('Training is ended!')

In [600]:
gpu_info = !nvidia-smi
gpu_info = '\n'.join(gpu_info)
print(gpu_info)

/bin/bash: line 1: nvidia-smi: command not found


In [601]:
CFG.wandb

True

In [602]:
main(model1)

test_accuracy,▁▁▁▁▁▁▁▁▁▁
test_loss,▁▁▁▁▁▁▁▁▁▁
train_accuracy,▁▁▁▁▁▁▁▁▁▁
train_loss,▁▂▆█▂▆▂▆▃▆
test_accuracy,0.76761
test_loss,0.66201
train_accuracy,0.72887
train_loss,0.6658



Epoch: 1


100%|██████████| 9/9 [00:00<00:00, 40.58it/s]


Train set: Average loss: 0.6763, Accuracy: 59%


Test set: Average loss: 0.6533, Accuracy: 53%

Epoch: 2


100%|██████████| 9/9 [00:00<00:00, 53.02it/s]


Train set: Average loss: 0.5680, Accuracy: 59%


Test set: Average loss: 0.6098, Accuracy: 53%

Epoch: 3


100%|██████████| 9/9 [00:00<00:00, 31.98it/s]


Train set: Average loss: 0.5477, Accuracy: 61%


Test set: Average loss: 0.5696, Accuracy: 65%

Epoch: 4


100%|██████████| 9/9 [00:00<00:00, 33.73it/s]


Train set: Average loss: 0.5335, Accuracy: 70%


Test set: Average loss: 0.5316, Accuracy: 75%

Epoch: 5


100%|██████████| 9/9 [00:00<00:00, 53.92it/s]


Train set: Average loss: 0.4902, Accuracy: 78%


Test set: Average loss: 0.4951, Accuracy: 85%

Epoch: 6


100%|██████████| 9/9 [00:00<00:00, 50.42it/s]


Train set: Average loss: 0.4837, Accuracy: 86%


Test set: Average loss: 0.4604, Accuracy: 87%

Epoch: 7


100%|██████████| 9/9 [00:00<00:00, 40.88it/s]


Train set: Average loss: 0.4195, Accuracy: 90%


Test set: Average loss: 0.4276, Accuracy: 91%

Epoch: 8


100%|██████████| 9/9 [00:00<00:00, 54.81it/s]


Train set: Average loss: 0.4064, Accuracy: 93%


Test set: Average loss: 0.3967, Accuracy: 92%

Epoch: 9


100%|██████████| 9/9 [00:00<00:00, 62.18it/s]


Train set: Average loss: 0.3924, Accuracy: 93%


Test set: Average loss: 0.3679, Accuracy: 94%

Epoch: 10


100%|██████████| 9/9 [00:00<00:00, 52.45it/s]


Train set: Average loss: 0.3605, Accuracy: 95%


Test set: Average loss: 0.3412, Accuracy: 94%
Training is ended!


#### Обучаем вторую модель

In [ ]:
#строим вторую модель
#документация, которую использовали https://docs.pytorch.org/docs/2.12/nn.html
class second_net(nn.Module): # наследуемся от класса nn.Module
    def __init__(self):
        super(second_net,self).__init__()
        hidden_1 = 256
        hidden_2 = 128

        self.net = nn.Sequential(
            nn.Linear(23, hidden_1),
            nn.BatchNorm1d(hidden_1),
            nn.ReLU(),
            nn.Dropout(0.1),

            nn.Linear(hidden_1, hidden_2),
            nn.BatchNorm1d(hidden_2),
            nn.ReLU(),
            nn.Dropout(0.1),

            nn.Linear(hidden_2, 2)
        )
    def forward(self,x):
        x = self.net(x)
        return x

In [ ]:
model = second_net()
#model = model.cuda() # переводим модель на GPU
print(model) # посмотрим на нашу модель

second_net(
  (net): Sequential(
    (0): Linear(in_features=23, out_features=256, bias=True)
    (1): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU()
    (3): Dropout(p=0.1, inplace=False)
    (4): Linear(in_features=256, out_features=128, bias=True)
    (5): BatchNorm1d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (6): ReLU()
    (7): Dropout(p=0.1, inplace=False)
    (8): Linear(in_features=128, out_features=2, bias=True)
  )
)


In [ ]:
# функция потерь (cross-entropy)
criterion = nn.CrossEntropyLoss()

# SGD (классический) со скоростью обучения (learning rate) = 0.01
optimizer = torch.optim.SGD(model.parameters(),lr = 0.01)

In [ ]:
main(model)

test_accuracy,▅▁▅▅█▅▁▅▅▅
test_loss,▃█▆▅▆█▄▃▃▁
train_accuracy,▃▄▂▇▄▄▄█▆▁
train_loss,▇▆▄▄▄█▆▁▄▆
test_accuracy,0.46479
test_loss,0.71593
train_accuracy,0.48944
train_loss,0.78704



Epoch: 1


100%|██████████| 9/9 [00:00<00:00, 104.37it/s]


Train set: Average loss: 0.6561, Accuracy: 69%


Test set: Average loss: 0.6424, Accuracy: 81%

Epoch: 2


100%|██████████| 9/9 [00:00<00:00, 100.44it/s]


Train set: Average loss: 0.5611, Accuracy: 71%


Test set: Average loss: 0.6177, Accuracy: 80%

Epoch: 3


100%|██████████| 9/9 [00:00<00:00, 101.64it/s]


Train set: Average loss: 0.5983, Accuracy: 74%


Test set: Average loss: 0.5993, Accuracy: 79%

Epoch: 4


100%|██████████| 9/9 [00:00<00:00, 90.67it/s]


Train set: Average loss: 0.5827, Accuracy: 71%


Test set: Average loss: 0.5894, Accuracy: 77%

Epoch: 5


100%|██████████| 9/9 [00:00<00:00, 104.86it/s]


Train set: Average loss: 0.5986, Accuracy: 71%


Test set: Average loss: 0.5853, Accuracy: 76%

Epoch: 6


100%|██████████| 9/9 [00:00<00:00, 106.43it/s]


Train set: Average loss: 0.6596, Accuracy: 69%


Test set: Average loss: 0.5835, Accuracy: 76%

Epoch: 7


100%|██████████| 9/9 [00:00<00:00, 101.48it/s]


Train set: Average loss: 0.5648, Accuracy: 70%


Test set: Average loss: 0.5826, Accuracy: 76%

Epoch: 8


100%|██████████| 9/9 [00:00<00:00, 97.79it/s]


Train set: Average loss: 0.5451, Accuracy: 70%


Test set: Average loss: 0.5812, Accuracy: 76%

Epoch: 9


100%|██████████| 9/9 [00:00<00:00, 100.57it/s]


Train set: Average loss: 0.5885, Accuracy: 70%


Test set: Average loss: 0.5813, Accuracy: 76%

Epoch: 10


100%|██████████| 9/9 [00:00<00:00, 102.08it/s]


Train set: Average loss: 0.5685, Accuracy: 69%


Test set: Average loss: 0.5815, Accuracy: 77%
Training is ended!


Снижение производительности модели во второй архитектуре связано с несколькими факторами:

1. Бачнорм добавляет дополнительную сложность модели и нормализует выводы каждого слоя, что может привести к потере важных признаков, учитывая, что в нашем датасете не так много признаков

2. Дропаут также может отсеивать часть признаков, а так как у нас мало приизнаков в датасете влияние может быть существенным.


В контексте медицинской диагностики каждый признак из датасета может быть очень важным, поэтому излишняя регуляризация на небольшом колчестве признаков может уменьшить качество модели

Итоговые показатели второй модели accuracy 76% (взяли модель с меньшим функционалом ошибки, а не с большим accuracy, так как функционал ошибки довольно сильно отличается, а значит модель с большим accuracy скорее всего чаще не находит положительные тесты пневмонии)

### Обучим третью модель

У нас есть несколько признаков, которые хорошо коррелируют с целевой переменной, поэтому если научить модель выбирать наилучшие признаки при помощи механизма внимания, это позволит улучшить ее качество, так как она будет лучше обучаться на более важном признаке.

In [43]:
# основная функция для экспериментов
def main(model):
    # настраиваем окружение для wandb
    if CFG.wandb:
        os.environ["WANDB_API_KEY"] = CFG.api
        wandb.init(project=CFG.project, entity=CFG.entity, reinit=True, config=class2dict(CFG))

    use_cuda = torch.cuda.is_available() # проверяем доступность GPU

    seed_everything(CFG.seed) # фиксируем сиды

    device = torch.device("cuda" if use_cuda else "cpu") # выделили устройство

    if CFG.wandb:
        wandb.watch(model, log='all') # логируем все (метрики, лоссы, градиенты)

    for epoch in range(1, 40): # цикл на эпохи
        print('\nEpoch:', epoch)
        train(model, device, train_loader, optimizer, criterion, epoch, CFG.wandb)
        test(model, device, test_loader, criterion, CFG.wandb)
    print('Training is ended!')

In [48]:
# строим третью модель с механизмом внимания
#документация про механизм внимания https://habr.com/ru/articles/1020624/
#в качестве функции softmax из статьи взяли функцию тангенса
class third_net(nn.Module):
    def __init__(self):
        super(third_net, self).__init__()
        self.attention = nn.Sequential(
            nn.Linear(23, 23),
            nn.Tanh(),
            nn.Linear(23, 23)
        )
        self.fc = nn.Sequential(
            nn.Linear(23, 128),
            nn.ReLU(),
            nn.Linear(128, 64),
            nn.ReLU(),
            nn.Linear(64, 2)
        )

    def forward(self, x):
        attn_weights = torch.softmax(self.attention(x), dim=1)
        weighted_x = x * attn_weights
        return self.fc(weighted_x)

In [49]:
model = third_net()
print(model)

third_net(
  (attention): Sequential(
    (0): Linear(in_features=23, out_features=23, bias=True)
    (1): Tanh()
    (2): Linear(in_features=23, out_features=1, bias=True)
  )
  (fc): Sequential(
    (0): Linear(in_features=23, out_features=128, bias=True)
    (1): ReLU()
    (2): Linear(in_features=128, out_features=64, bias=True)
    (3): ReLU()
    (4): Linear(in_features=64, out_features=2, bias=True)
  )
)


In [50]:
# функция потерь (cross-entropy)
criterion = nn.CrossEntropyLoss()

# SGD (классический) со скоростью обучения (learning rate) = 0.01
optimizer = torch.optim.SGD(model.parameters(),lr = 0.01)

In [51]:
main(model)

test_accuracy,▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
test_loss,█▇▆▆▅▄▄▃▃▂▂▂▂▁▁▁
train_accuracy,▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
train_loss,▅▅▃▅▃▄▄▅▃▂▆▁█▃▃▅▆
test_accuracy,0.57042
test_loss,0.68336
train_accuracy,0.5757
train_loss,0.69388



Epoch: 1


100%|██████████| 9/9 [00:00<00:00, 89.91it/s]


Train set: Average loss: 0.6884, Accuracy: 45%


Test set: Average loss: 0.6862, Accuracy: 47%

Epoch: 2


100%|██████████| 9/9 [00:00<00:00, 113.71it/s]


Train set: Average loss: 0.6702, Accuracy: 54%


Test set: Average loss: 0.6717, Accuracy: 52%

Epoch: 3


100%|██████████| 9/9 [00:00<00:00, 112.11it/s]


Train set: Average loss: 0.6462, Accuracy: 58%


Test set: Average loss: 0.6579, Accuracy: 58%

Epoch: 4


100%|██████████| 9/9 [00:00<00:00, 109.25it/s]


Train set: Average loss: 0.6408, Accuracy: 60%


Test set: Average loss: 0.6443, Accuracy: 61%

Epoch: 5


100%|██████████| 9/9 [00:00<00:00, 115.20it/s]


Train set: Average loss: 0.6273, Accuracy: 63%


Test set: Average loss: 0.6307, Accuracy: 63%

Epoch: 6


100%|██████████| 9/9 [00:00<00:00, 121.33it/s]


Train set: Average loss: 0.6107, Accuracy: 65%


Test set: Average loss: 0.6168, Accuracy: 72%

Epoch: 7


100%|██████████| 9/9 [00:00<00:00, 113.72it/s]


Train set: Average loss: 0.6019, Accuracy: 69%


Test set: Average loss: 0.6024, Accuracy: 75%

Epoch: 8


100%|██████████| 9/9 [00:00<00:00, 99.43it/s]


Train set: Average loss: 0.5986, Accuracy: 72%


Test set: Average loss: 0.5875, Accuracy: 77%

Epoch: 9


100%|██████████| 9/9 [00:00<00:00, 119.02it/s]


Train set: Average loss: 0.5833, Accuracy: 75%


Test set: Average loss: 0.5719, Accuracy: 79%

Epoch: 10


100%|██████████| 9/9 [00:00<00:00, 122.43it/s]


Train set: Average loss: 0.5441, Accuracy: 79%


Test set: Average loss: 0.5556, Accuracy: 84%

Epoch: 11


100%|██████████| 9/9 [00:00<00:00, 110.10it/s]


Train set: Average loss: 0.5533, Accuracy: 82%


Test set: Average loss: 0.5385, Accuracy: 87%

Epoch: 12


100%|██████████| 9/9 [00:00<00:00, 119.17it/s]


Train set: Average loss: 0.5230, Accuracy: 84%


Test set: Average loss: 0.5208, Accuracy: 90%

Epoch: 13


100%|██████████| 9/9 [00:00<00:00, 110.69it/s]


Train set: Average loss: 0.5421, Accuracy: 86%


Test set: Average loss: 0.5022, Accuracy: 91%

Epoch: 14


100%|██████████| 9/9 [00:00<00:00, 121.83it/s]


Train set: Average loss: 0.5244, Accuracy: 89%


Test set: Average loss: 0.4829, Accuracy: 92%

Epoch: 15


100%|██████████| 9/9 [00:00<00:00, 105.98it/s]


Train set: Average loss: 0.4752, Accuracy: 90%


Test set: Average loss: 0.4632, Accuracy: 94%

Epoch: 16


100%|██████████| 9/9 [00:00<00:00, 118.01it/s]


Train set: Average loss: 0.4374, Accuracy: 90%


Test set: Average loss: 0.4430, Accuracy: 94%

Epoch: 17


100%|██████████| 9/9 [00:00<00:00, 107.78it/s]


Train set: Average loss: 0.4530, Accuracy: 92%


Test set: Average loss: 0.4226, Accuracy: 95%

Epoch: 18


100%|██████████| 9/9 [00:00<00:00, 116.30it/s]


Train set: Average loss: 0.4468, Accuracy: 93%


Test set: Average loss: 0.4022, Accuracy: 95%

Epoch: 19


100%|██████████| 9/9 [00:00<00:00, 104.51it/s]


Train set: Average loss: 0.4262, Accuracy: 94%


Test set: Average loss: 0.3821, Accuracy: 96%

Epoch: 20


100%|██████████| 9/9 [00:00<00:00, 110.77it/s]


Train set: Average loss: 0.3833, Accuracy: 94%


Test set: Average loss: 0.3624, Accuracy: 96%

Epoch: 21


100%|██████████| 9/9 [00:00<00:00, 97.99it/s]


Train set: Average loss: 0.3889, Accuracy: 95%


Test set: Average loss: 0.3433, Accuracy: 97%

Epoch: 22


100%|██████████| 9/9 [00:00<00:00, 98.58it/s]


Train set: Average loss: 0.3494, Accuracy: 95%


Test set: Average loss: 0.3249, Accuracy: 98%

Epoch: 23


100%|██████████| 9/9 [00:00<00:00, 113.34it/s]


Train set: Average loss: 0.3489, Accuracy: 96%


Test set: Average loss: 0.3074, Accuracy: 99%

Epoch: 24


100%|██████████| 9/9 [00:00<00:00, 101.82it/s]


Train set: Average loss: 0.3012, Accuracy: 96%


Test set: Average loss: 0.2908, Accuracy: 98%

Epoch: 25


100%|██████████| 9/9 [00:00<00:00, 110.17it/s]


Train set: Average loss: 0.3164, Accuracy: 96%


Test set: Average loss: 0.2752, Accuracy: 97%

Epoch: 26


100%|██████████| 9/9 [00:00<00:00, 127.24it/s]


Train set: Average loss: 0.2434, Accuracy: 96%


Test set: Average loss: 0.2606, Accuracy: 97%

Epoch: 27


100%|██████████| 9/9 [00:00<00:00, 131.86it/s]


Train set: Average loss: 0.3191, Accuracy: 96%


Test set: Average loss: 0.2469, Accuracy: 97%

Epoch: 28


100%|██████████| 9/9 [00:00<00:00, 114.39it/s]


Train set: Average loss: 0.2325, Accuracy: 96%


Test set: Average loss: 0.2342, Accuracy: 97%

Epoch: 29


100%|██████████| 9/9 [00:00<00:00, 122.90it/s]


Train set: Average loss: 0.2384, Accuracy: 96%


Test set: Average loss: 0.2224, Accuracy: 98%

Epoch: 30


100%|██████████| 9/9 [00:00<00:00, 115.72it/s]


Train set: Average loss: 0.2158, Accuracy: 96%


Test set: Average loss: 0.2114, Accuracy: 98%

Epoch: 31


100%|██████████| 9/9 [00:00<00:00, 117.87it/s]


Train set: Average loss: 0.1605, Accuracy: 96%


Test set: Average loss: 0.2013, Accuracy: 98%

Epoch: 32


100%|██████████| 9/9 [00:00<00:00, 114.15it/s]


Train set: Average loss: 0.1854, Accuracy: 96%


Test set: Average loss: 0.1920, Accuracy: 98%

Epoch: 33


100%|██████████| 9/9 [00:00<00:00, 114.32it/s]


Train set: Average loss: 0.2040, Accuracy: 96%


Test set: Average loss: 0.1833, Accuracy: 98%

Epoch: 34


100%|██████████| 9/9 [00:00<00:00, 99.19it/s]


Train set: Average loss: 0.2143, Accuracy: 96%


Test set: Average loss: 0.1753, Accuracy: 98%

Epoch: 35


100%|██████████| 9/9 [00:00<00:00, 117.29it/s]


Train set: Average loss: 0.2056, Accuracy: 96%


Test set: Average loss: 0.1679, Accuracy: 98%

Epoch: 36


100%|██████████| 9/9 [00:00<00:00, 114.56it/s]


Train set: Average loss: 0.1614, Accuracy: 96%


Test set: Average loss: 0.1610, Accuracy: 97%

Epoch: 37


100%|██████████| 9/9 [00:00<00:00, 116.07it/s]


Train set: Average loss: 0.2101, Accuracy: 96%


Test set: Average loss: 0.1547, Accuracy: 97%

Epoch: 38


100%|██████████| 9/9 [00:00<00:00, 113.43it/s]


Train set: Average loss: 0.1498, Accuracy: 97%


Test set: Average loss: 0.1487, Accuracy: 97%

Epoch: 39


100%|██████████| 9/9 [00:00<00:00, 81.60it/s]


Train set: Average loss: 0.1993, Accuracy: 97%


Test set: Average loss: 0.1432, Accuracy: 97%
Training is ended!


Увеличили количество эпох, так как заметили постепенное уменьшение ошибки с каждой эпохой без переобучения. По итогу получили accuracy 97% с уровнем ошибки 0.1432